### PR TITLE
client: emit more information in errors

### PR DESCRIPTION
**Pull request by daviddrysdale**
Probably best reviewed commit-by-commit as I refactored various of the client tests along the way.

For #89

### CODE REVIEW

File: client/logclient_test.go [hunk 1]
```diff
@@ -176,9 +176,9 @@ func TestGetEntriesErrors(t *testing.T) {
 		{start: 1, end: 2, rsp: "", wantErr: "EOF"},
 		{start: 0, end: -1, wantErr: "end should be >= 0"},
 		{start: 3, end: 2, wantErr: "start should be <= end"},
-		{start: 1, end: 2, rsp: "not-json", wantErr: "invalid"},
-		{start: 1, end: 2, rsp: `{"entries":[{"leaf_input":"bogus","extra_data":"bogus"}]}`, wantErr: "illegal base64"},
-		{start: 1, end: 2, rsp: `{"entries":[{"leaf_input":"bbbb","extra_data":"bbbb"}]}`, wantErr: "failed to unmarshal"},
+		{start: 4, end: 5, rsp: "not-json", wantErr: "invalid"},
+		{start: 5, end: 6, rsp: `{"entries":[{"leaf_input":"bogus","extra_data":"bogus"}]}`, wantErr: "illegal base64"},
```
Comment (phad) [inline client/logclient_test.go]: Not sure why these start/end values changed?  Was that to trigger some different behaviours in the test?

Comment (daviddrysdale) [reply client/logclient_test.go]: I changed it so every test case uses different values, so it's easier to spot which one failed. I could have added a description field...but I didn't.

File: client/logclient_test.go [hunk 2]
```diff
@@ -164,6 +165,39 @@ func TestGetEntries(t *testing.T) {
 	}
 }
 
+func TestGetEntriesErrors(t *testing.T) {
+	ctx := context.Background()
+	var tests = []struct {
+		start   int64
```
[the remark this answers is not in the document]
Comment (daviddrysdale) [reply client/logclient_test.go]: Done.

File: client/logclient_test.go [hunk 3]
```diff
@@ -164,6 +165,39 @@ func TestGetEntries(t *testing.T) {
...
+		end     int64
+		rsp     string
+		wantErr string
```
[the remark this answers is not in the document]
Comment (daviddrysdale) [reply client/logclient_test.go]: Done.

File: client/logclient_test.go [hunk 4]
```diff
@@ -164,6 +165,39 @@ func TestGetEntries(t *testing.T) {
...
+	}{
+		{start: 1, end: 2, rsp: "", wantErr: "EOF"},
+		{start: 0, end: -1, wantErr: "end should be >= 0"},
+		{start: 3, end: 2, wantErr: "start should be <= end"},
+		{start: 1, end: 2, rsp: "not-json", wantErr: "invalid"},
+		{start: 1, end: 2, rsp: `{"entries":[{"leaf_input":"bogus","extra_data":"bogus"}]}`, wantErr: "illegal base64"},
+		{start: 1, end: 2, rsp: `{"entries":[{"leaf_input":"bbbb","extra_data":"bbbb"}]}`, wantErr: "failed to unmarshal"},
```
[the remark this answers is not in the document]
Comment (daviddrysdale) [reply client/logclient_test.go]: Pre-existing behaviour, so feel free to raise an issue if you think it should be changed.

File: client/logclient_test.go [hunk 5]
```diff
@@ -164,6 +165,39 @@ func TestGetEntries(t *testing.T) {
...
+	}
+
+	for _, test := range tests {
+		ts := serveRspAt(t, "/ct/v1/get-entries", test.rsp)
+		defer ts.Close()
+		client, err := New(ts.URL, &http.Client{}, jsonclient.Options{})
+		if err != nil {
+			t.Errorf("Failed to create client: %v", err)
```
[the remark this answers is not in the document]
Comment (daviddrysdale) [reply client/logclient_test.go]: This is a `t.Errorf` not a `fmt.Errorf`

File: client/logclient_test.go [hunk 6]
```diff
@@ -164,6 +165,39 @@ func TestGetEntries(t *testing.T) {
...
+			continue
+		}
+		got, err := client.GetEntries(ctx, test.start, test.end)
+		if err == nil {
+			t.Errorf("GetEntries(%d, %d)=%+v, nil; want nil, %q", test.start, test.end, got, test.wantErr)
```
[the remark this answers is not in the document]
Comment (daviddrysdale) [reply client/logclient_test.go]: Done.

File: client/logclient_test.go [hunk 7]
```diff
@@ -164,6 +165,39 @@ func TestGetEntries(t *testing.T) {
...
+		} else if !strings.Contains(err.Error(), test.wantErr) {
```
[the remark this answers is not in the document]
Comment (daviddrysdale) [reply client/logclient_test.go]: Now falls through to a check of `got` (and besides, doing that would be 1 line longer).

File: client/logclient_test.go [hunk 8]
```diff
@@ -94,66 +94,63 @@ func b64(s string) []byte {
...
-			if err != nil {
-				t.Fatalf("Failed to json-marshal test precertificate proof: %v", err)
-			}
-			w.Write(data)
-		case r.URL.Path == "/ct/v1/add-json":
-			w.Write([]byte(AddJSONResp))
-		default:
+// serveHandlerAt returns a test HTTP server that only expects requests at the given path, and invokes
+// the provided handler for that path.
+func serveHandlerAt(t *testing.T, path string, handler func(http.ResponseWriter, *http.Request)) *httptest.Server {
```
[the remark this answers is not in the document]
Comment (daviddrysdale) [reply client/logclient_test.go]: Added an overall TODO

File: client/logclient_test.go [hunk 9]
```diff
@@ -94,66 +94,63 @@ func b64(s string) []byte {
...
+		if r.URL.Path == path {
+			handler(w, r)
+		} else {
 			t.Fatalf("Incorrect URL path: %s", r.URL.Path)
 		}
 	}))
-	return hs
+}
+
+// serveRspAt returns a test HTTP server that returns a canned response for a given path.
```
[the remark this answers is not in the document]
Comment (daviddrysdale) [reply client/logclient_test.go]: TBH seems overkill for a comment on an unexported function in a test package

File: client/logclient_test.go [hunk 10]
```diff
@@ -94,66 +94,63 @@ func b64(s string) []byte {
...
+func serveRspAt(t *testing.T, path string, rsp string) *httptest.Server {
```
[the remark this answers is not in the document]
Comment (daviddrysdale) [reply client/logclient_test.go]: I've got long history of using `req` / `rsp` as a pair.

[the remark this answers is not in the document]
Comment (daviddrysdale) [reply client/logclient_test.go]: Done

File: client/logclient_test.go [hunk 11]
```diff
@@ -94,66 +94,63 @@ func b64(s string) []byte {
...
+	return serveHandlerAt(t, path, func(w http.ResponseWriter, r *http.Request) {
+		fmt.Fprintf(w, rsp)
+	})
+}
+
+func sctToJSON(rawSCT []byte) ([]byte, error) {
+	var sct ct.SignedCertificateTimestamp
+	_, err := tls.Unmarshal(rawSCT, &sct)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to tls-unmarshal test certificate proof: %v", err)
```
[the remark this answers is not in the document]
Comment (daviddrysdale) [reply client/logclient_test.go]: Done

File: client/logclient_test.go [hunk 12]
```diff
@@ -94,66 +94,63 @@ func b64(s string) []byte {
...
+	}
+	data, err := json.Marshal(sct)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to json-marshal test certificate proof: %v", err)
```
[the remark this answers is not in the document]
Comment (daviddrysdale) [reply client/logclient_test.go]: Done

File: client/logclient_test.go [hunk 13]
```diff
@@ -432,6 +497,37 @@ func TestGetSTHConsistency(t *testing.T) {
 	}
 }
 
+func TestGetSTHConsistencyErrors(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (daviddrysdale) [reply client/logclient_test.go]: I don't understand this comment

File: client/logclient_test.go [hunk 14]
```diff
@@ -207,10 +209,13 @@ func TestGetSTH(t *testing.T) {
 }
 
 func TestAddChainRetries(t *testing.T) {
+	if testing.Short() {
```
[the remark this answers is not in the document]
Comment (daviddrysdale) [reply client/logclient_test.go]: It isn't.  It's just useful to be able to skip the real-time retry test if someone is (manually) running with `-test.short`.

File: client/multilog_test.go [hunk 15]
```diff
@@ -333,7 +335,18 @@ func TestIndexByDate(t *testing.T) {
 }
 
 func TestTemporalAddChain(t *testing.T) {
-	hs := ctServer(t)
+	hs := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
```
[the remark this answers is not in the document]
Comment (daviddrysdale) [reply client/multilog_test.go]: `serveSctAt` only copes with a single entrypoint.

File: jsonclient/client.go [hunk 16]
```diff
@@ -121,12 +121,14 @@ func New(uri string, hc *http.Client, opts Options) (*JSONClient, error) {
...
-// the response as a JSON representation of the rsp structure.  The provided
-// context is used to control the HTTP call.
-func (c *JSONClient) GetAndParse(ctx context.Context, path string, params map[string]string, rsp interface{}) (*http.Response, error) {
+// GetAndParse makes a HTTP GET call to the given path, and attempta to parse
+// the response as a JSON representation of the rsp structure.  Returns the
+// http.Response, the body of the response, and an error.  Note that the
+// returned http.Response can be non-nil even when an error is returned,
+// in particular when the HTTP status is not OK or when the JSON parsing fails.
+func (c *JSONClient) GetAndParse(ctx context.Context, path string, params map[string]string, rsp interface{}) (*http.Response, []byte, error) {
 	if ctx == nil {
```
[the remark this answers is not in the document]
Comment (daviddrysdale) [reply jsonclient/client.go]: This is a hold-over from an API change a while ago, where we added the context parameters -- at the time it seemed more helpful for the transition.

File: jsonclient/client_test.go [hunk 17]
```diff
@@ -221,12 +223,15 @@ func TestGetAndParse(t *testing.T) {
 
 	for _, test := range tests {
 		var result TestStruct
-		httpRsp, err := logClient.GetAndParse(ctx, test.uri, test.params, &result)
+		httpRsp, body, err := logClient.GetAndParse(ctx, test.uri, test.params, &result)
+		if gotBody := (body != nil); gotBody != test.wantBody {
+			t.Errorf("GetAndParse(%q) returned body %v, wanted %v", test.uri, gotBody, test.wantBody)
```
[the remark this answers is not in the document]
Comment (daviddrysdale) [reply jsonclient/client_test.go]: Done.

File: client/logclient_test.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (phad) [reply client/logclient_test.go]: If it saves me having to go and read a bunch of code then why not?

File: client/logclient_test.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (phad) [reply client/logclient_test.go]: I mean apply the same changes I've requested above to the other similar functions in this file.

[the remark this answers is not in the document]
Comment (daviddrysdale) [reply client/logclient_test.go]: Ah, 'And' not 'Add', makes sense now.

**Pull request merged**